python/pyspark/sql/tests.py

```diff
@@ -3274,6 +3274,39 @@ def test_vectorized_udf_mix_udf(self):
                 df.select(row_by_row_udf(col('id')), pd_udf(col('id'))).collect()
 
 
+@unittest.skipIf(not _have_pandas or not _have_arrow, "Pandas or Arrow not installed")
+class VectorizedUDAFTests(ReusedPySparkTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        ReusedPySparkTestCase.setUpClass()
+        cls.spark = SparkSession(cls.sc)
+
+    @classmethod
+    def tearDownClass(cls):
+        ReusedPySparkTestCase.tearDownClass()
+        cls.spark.stop()
+
+    def test_vectorized_udaf_basic(self):
+        df = self.spark.range(100).select(col('id').alias('n'), (col('id') % 2 == 0).alias('g'))
+
+        @pandas_udaf(LongType(), algebraic=True)
+        def p_sum(v):
+            return v.sum()
+
+        @pandas_udaf(
+            DoubleType(),
+            algebraic=False,
+            partial=lambda v: (v.sum(), len(v)),
+            bufferType=StructType().add("sum", LongType()).add("count", LongType()))
```

Comment:
I'm a little worried about the API here. I think the vectorized python UDF/UDAF is mostly for reusing pandas functions, other than performance. For better performance, users should implement UDF/UDAF in Scala/Java, register and use it in Python.

Talking about reusing pandas functions, we should make the API as simple as possible, my proposal:
```
@pandas_udaf(LongType(), supportsPartial=True)
def p_sum(v):
  return v.sum()
```

If the pandas function doesn't support partial, we should just plan a single stage aggregate.